src/ir/opt/cfold.rs

```diff
@@ -244,6 +244,29 @@ pub fn fold_cache(node: &Term, cache: &mut TermCache<Term>) -> Term {
                     b.width() + w,
                 ))))
             }),
+            Op::Map(op) => match (get(0).as_array_opt(), get(1).as_array_opt()) {
+                (Some(a), Some(b)) => {
+                    // TODO: extend for n-ary arrays
+                    let mut res = a.clone();
+                    let mut merge = ArrayMerge::new(a.clone(), b.clone());
+                    for (i, va, vb) in merge.into_iter() {
+                        let r = fold_cache(
+                            &term![*op.clone(); leaf_term(Op::Const(va.clone())), leaf_term(Op::Const(vb.clone()))],
+                            cache,
+                        );
+                        match r.as_value_opt() {
+                            Some(v) => {
+                                res = res.clone().store(i, v.clone());
```

Comment:
This line clones `res` on every pass through the loop. That is quadratic time for totally reasonable values of `a` and `b`. And then the previous copy just gets thrown away!

```suggestion
                                res = res.store(i, v.clone());
```